python/taichi/lang/kernel.py

```diff
@@ -456,7 +456,9 @@ def __call__(self, *args, **kwargs):
         instance_id, arg_features = self.mapper.lookup(args)
         key = (self.func, instance_id)
         self.materialize(key=key, args=args, arg_features=arg_features)
-        return self.compiled_functions[key](*args)
+        ret = self.compiled_functions[key](*args)
+        print(taichi_lang_core.pop_python_print_buffer(), end='')
```

Comment:
I don't think this works for all platforms, does it? See https://github.com/taichi-dev/taichi/blob/37976e750e1579ff9328f3d25f1a63cb3b3292e2/tests/python/test_print.py#L17-L19. According to that comment, CUDA's print buffer isn't flushed until `ti.sync()` is called. Maybe we need to fix the CUDA backend first?

Reply:
How about this: move this line to ti.sync() and doc this behavior, so that we get the same behavior along backends.

Reply:
Yep SGTM. (At one point I think `ti.sync` should only do one thing, but it also feels reasonable to flush the print buffer at synchronization...)